### PR TITLE
drivers: tach: microchip: Add MEC172x TACH support

### PR DESCRIPTION
**Pull request by jvasanth1**
Add support for MEC172x TACH and pin control changes.

### CODE REVIEW

File: drivers/sensor/mchp_tach_xec/tach_mchp_xec.c [hunk 1]
```diff
@@ -116,19 +133,25 @@ static const struct sensor_driver_api tach_xec_driver_api = {
 	.channel_get = tach_xec_channel_get,
 };
 
+#define XEC_TACH_CONFIG(inst)						\
+	static const struct tach_xec_config tach_xec_config_##inst = {	\
+		.regs = NULL,						\
```
Comment (MaureenHelm) [inline drivers/sensor/mchp_tach_xec/tach_mchp_xec.c:155]: This can't be right

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/mchp_tach_xec/tach_mchp_xec.c (hunk 1 above)
Comment (jvasanth1) [reply drivers/sensor/mchp_tach_xec/tach_mchp_xec.c]: Fixed

File: drivers/sensor/mchp_tach_xec/tach_mchp_xec.c [hunk 2]
```diff
@@ -99,9 +97,28 @@ static int tach_xec_channel_get(const struct device *dev,
 	return 0;
 }
 
+static void tach_xec_sleep_clr(const struct device *dev)
+{
+	const struct tach_xec_config * const cfg = dev->config;
+	struct pcr_regs * const pcr = (struct pcr_regs * const)(
+		DT_REG_ADDR_BY_IDX(DT_NODELABEL(pcr), 0));
```
Comment (gmarull) [inline drivers/sensor/mchp_tach_xec/tach_mchp_xec.c:110]: If your device depends on another, the best thing is to specify the dependency explicitely in Devicetree. For example, by doing:


```
&tach {
   ...
    pcr = <&pcr>;
};
```

and then collect it's address (or instance if there's a proper driver) in the config. You could then get the address by doing something like

```
.pcr_regs = (struct pcr_regs *)DT_REG_ADDR(DT_INST_PHANDLE(n, pcr));
```

However, it looks like PCR is a clock/reset controller, for which Zephyr has a better way to solve the problem (a reset API was recently introduced for example). For clocks, let's take for example STM32:

```
spi3: ... {
    ...    
    clocks = <&rcc STM32_CLOCK_BUS_APB1 0x0...>;
};
```

As you can see, we have a clock cell that specifies both the controller (pcr in your case, rcc for STM32) and then clock properties for that particular peripheral.

@erwango could you provide some feedback here? Should we continue accepting custom stuff like this when we have a better way?

**Comment by albertofloyd** [comment]
@jvasanth1 I understand this is not a driver that should be enabled for minimum config.
However, I'm expecting when a particular app does enable it will be compiled. This is not the case 
 i.e.  west build -c -b mec172xevb_assy6906 -- --DCONFIG_SENSOR=y

**Comment by jvasanth1** [comment]
> @jvasanth1 I understand this is not a driver that should be enabled for minimum config. However, I'm expecting when a particular app does enable it will be compiled. This is not the case i.e. west build -c -b mec172xevb_assy6906 -- --DCONFIG_SENSOR=y

Need to enable CONFIG_TACH_XEC=y as well

**Review by albertofloyd — CHANGES_REQUESTED** [review]
Need to fix Kconfig dependencies

**Comment by albertofloyd** [comment]
> > @jvasanth1 I understand this is not a driver that should be enabled for minimum config. However, I'm expecting when a particular app does enable it will be compiled. This is not the case i.e. west build -c -b mec172xevb_assy6906 -- --DCONFIG_SENSOR=y
> 
> Need to enable CONFIG_TACH_XEC=y as well

No, that should be automatically enabled, see the same for MEC152x and other SoC drivers

**Comment by jvasanth1** [comment]
> > > @jvasanth1 I understand this is not a driver that should be enabled for minimum config. However, I'm expecting when a particular app does enable it will be compiled. This is not the case i.e. west build -c -b mec172xevb_assy6906 -- --DCONFIG_SENSOR=y
> > 
> > 
> > Need to enable CONFIG_TACH_XEC=y as well
> 
> No, that should be automatically enabled, see the same for MEC152x and other SoC drivers

Updated accordingly.

File: drivers/sensor/mchp_tach_xec/tach_mchp_xec.c (hunk 2 above)
Comment (erwango) [reply drivers/sensor/mchp_tach_xec/tach_mchp_xec.c:110]: > @erwango could you provide some feedback here? Should we continue accepting custom stuff like this when we have a better way?

Tbh, on STM32 we're using same shortcut for clocks as clock_controller is a singleton.
This being said, I agree that at least the bindings should be "canonical": <&phandle cell1 cell2>;

Comment (jvasanth1) [reply drivers/sensor/mchp_tach_xec/tach_mchp_xec.c:110]: We will consider this for future drivers, don't want to change at this point.

Comment (MaureenHelm) [reply drivers/sensor/mchp_tach_xec/tach_mchp_xec.c:110]: We don't need to block this PR on this point, but current drivers do need to be changed too.

**Review by MaureenHelm — APPROVED** [review]

**Review by albertofloyd — APPROVED** [review]

**Pull request merged**